### PR TITLE
WaitFor: Add health checks for Qdrant

### PR DESCRIPTION
**Pull request by Alirexaa**
## Description
Adds `WaitFor` support for Qdrant

Related: #5645

## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [X] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [X] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5767)

### CODE REVIEW

File: src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs [hunk 1]
```diff
@@ -117,4 +143,46 @@ public static IResourceBuilder<TDestination> WithReference<TDestination>(this IR
...
+            }
+
+            if (connectionBuilder.ContainsKey("Key"))
+            {
+                key = connectionBuilder["Key"].ToString();
+            }
+        }
+
+        var factory = sp.GetRequiredService<IHttpClientFactory>();
+        var client = factory.CreateClient();
```
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:180]: Even though this health check is in the components path, it looks like its only used in the hosting package. Given this we assert a lot more control over the nature of the endpoint we are connecting to in the hosting package.

I agree that we might want to make this a bit more robust if we were registering this when we add it to `Aspire.Qdrant.Client` (we probably do want to get there).

[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:180]: I filed #5768 to add health checks to `Aspire.Qdrant.Client`.

File: src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs [hunk 2]
```diff
@@ -117,4 +143,46 @@ public static IResourceBuilder<TDestination> WithReference<TDestination>(this IR
...
+        client.BaseAddress = endpoint;
+        if (key is not null)
```
[the remark this answers is not in the document]
Comment (Alirexaa) [reply src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:182]: I used named clients, so we don't need this.

**Comment by sebastienros** [comment]
Digging a little more into the `HttpClient` usage here, I wonder if it would be better to resolve it in `CheckHealthAsync` instead and dispose it in this same method to follow a better `IHttpClientFactory` pattern instead of keeping a long-lived instance around. And then the named client could be configured on registration like this:

```c#
builder.Services.AddHttpClient(
    "qdrant-healthchecks",
    client =>
    {
        // Set the base address of the named client.
        client.BaseAddress = ...;

        client.DefaultRequestHeaders.Add("Api-Key", ...)
    });
```

**Comment by Alirexaa** [comment]
> Digging a little more into the `HttpClient` usage here, I wonder if it would be better to resolve it in `CheckHealthAsync` instead and dispose it in this same method to follow a better `IHttpClientFactory` pattern instead of keeping a long-lived instance around. And then the named client could be configured on registration like this:
> 
> ```cs
> builder.Services.AddHttpClient(
>     "qdrant-healthchecks",
>     client =>
>     {
>         // Set the base address of the named client.
>         client.BaseAddress = ...;
> 
>         client.DefaultRequestHeaders.Add("Api-Key", ...)
>     });
> ```

We haven't base  address before application start, so we can't use `ServiceCollection` to add named client.
We subscribe `ConnectionStringAvailableEvent` to get base address and in this step we can't add service in `ServiceCollection`

File: src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs [hunk 3]
```diff
@@ -42,6 +46,27 @@ public static IResourceBuilder<QdrantServerResource> AddQdrant(this IDistributed
         var apiKeyParameter = apiKey?.Resource ??
             ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-Key", special: false);
         var qdrant = new QdrantServerResource(name, apiKeyParameter);
+
+        builder.Services.AddHttpClient();
```
Comment (davidfowl) [inline src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:50]: I'm not convinced there's a benefit to using http client factory here.

**Comment by mitchdenny** [comment]
@Alirexaa we are looking to snap a branch for .NET Aspire 9.0 at the end of this week? Did you have a chance to look at @davidfowl's feedback on http factory?

File: src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs [hunk 4]
```diff
@@ -117,4 +143,46 @@ public static IResourceBuilder<TDestination> WithReference<TDestination>(this IR
...
+            }
+
+            if (connectionBuilder.TryGetValue("Key", out var keyValue))
+            {
+                key = keyValue.ToString();
+            }
+        }
+
+        var factory = sp.GetRequiredService<IHttpClientFactory>();
+        var client = factory.CreateClient("qdrant-healthchecks");
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Qdrant/QdrantBuilderExtensions.cs:180]: Please make sure that this works with multiple Qdrant instances.

**Comment by Alirexaa** [comment]
> @Alirexaa we are looking to snap a branch for .NET Aspire 9.0 at the end of this week? Did you have a chance to look at @davidfowl's feedback on http factory?

I will look.

**Comment by mitchdenny** [comment]
Replacing with #6044 with a few minor tweaks. I couldn't push to your branch for some reason.

**Pull request closed without merging**